### PR TITLE
MRRTF-167: (optional) filter out MCH ROFs not matching any ITS-selected IRFrame

### PR DESCRIPTION
**Pull request by aphecetche**
@aferrero2707 here's the introduction of the filter on IRFrames. When applied, it is done at the end of the [`TimeClusterFinderSpec.cxx`](https://github.com/AliceO2Group/AliceO2/compare/dev...aphecetche:mrrtf-167-filter-rof-using-its-irframes?expand=1#diff-a4c8993892ed009dcbbfbab877f0e1f7de716da869dc7c9d8e4ba0c6409570d1), like e.g. the trackable-only filter.

@pillot note that the trackable filter is now always doing something (https://github.com/AliceO2Group/AliceO2/pull/8833#discussion_r873669287) : the no-op was removed.

I did not test it with "real" `ITS/IRFRAMES` but with fake (and thus controlled) ones using a custom made test device :  [`fake-irframe-generator-workflow.cxx`](https://github.com/AliceO2Group/AliceO2/compare/dev...aphecetche:mrrtf-167-filter-rof-using-its-irframes?expand=1#diff-1e0ebb71dfa0d44aa07800d4c76c948a02f6b6fc2f9af7fcec31bd6199ff7bc6)

### CODE REVIEW

File: Detectors/MUON/MCH/ROFFiltering/include/MCHROFFiltering/MultiplicityFilter.h [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+// or submit itself to any jurisdiction.
+
+#ifndef O2_MCH_ROFFILTERING_MULTIPLICITY_FILTER_H_
+#define O2_MCH_ROFFILTERING_MULTIPLICITY_FILTER_H_
+
+#include <functional>
+#include "DataFormatsMCH/ROFRecord.h"
+#include <array>
+#include <gsl/span>
+#include "MCHBase/Trackable.h"
```
Comment (pillot) [inline Detectors/MUON/MCH/ROFFiltering/include/MCHROFFiltering/MultiplicityFilter.h]: the 3 includes above are not needed, are they?

File: Detectors/MUON/MCH/ROFFiltering/src/IRFrameFilter.cxx [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+namespace o2::mch
+{
+ROFFilter createIRFrameFilter(gsl::span<const o2::dataformats::IRFrame> irframes)
+{
+  return [irframes](const ROFRecord& rof) {
+    InteractionRecord rofStart{rof.getBCData()};
+    InteractionRecord rofEnd = rofStart + rof.getBCWidth();
+    IRFrame ref(rofStart, rofEnd);
+    for (const auto& ir : irframes) {
+      // LOGP(info, "TOTO ref {} vs ir {}", ref.asString(), ir.asString());
```
Comment (pillot) [inline Detectors/MUON/MCH/ROFFiltering/src/IRFrameFilter.cxx]: is this a leftover? Note that you can also comment (or remove) the Framework/Logger.h include if this is commented (or removed)

File: Detectors/MUON/MCH/ROFFiltering/src/fake-irframe-generator-workflow.cxx [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+#include <cstring>
+#include <fmt/format.h>
+#include <fstream>
+#include <iostream>
+#include <random>
+#include <stdexcept>
+#include <vector>
+#include <rapidjson/document.h>
+#include <rapidjson/istreamwrapper.h>
+
```
Comment (pillot) [inline Detectors/MUON/MCH/ROFFiltering/src/fake-irframe-generator-workflow.cxx:31]: some include like chrono or random are not needed, right?

File: Detectors/MUON/MCH/TimeClustering/src/TimeClusterFinderSpec.cxx [hunk 4]
```diff
@@ -160,18 +176,24 @@ class TimeClusterFinderTask
...
                            std::string_view inputDigitDataDescription,
                            std::string_view inputDigitRofDataDescription,
-                           std::string_view outputDigitRofDataDescription)
+                           std::string_view outputDigitRofDataDescription,
+                           std::string_view inputIRFrameDataDescription)
 {
   std::string input = fmt::format("rofs:MCH/{}/0;digits:MCH/{}/0",
                                   inputDigitRofDataDescription.data(),
                                   inputDigitDataDescription.data());
+  if (TimeClusterizerParam::Instance().irFramesOnly && inputIRFrameDataDescription.size()) {
```
Comment (pillot) [inline Detectors/MUON/MCH/TimeClustering/src/TimeClusterFinderSpec.cxx:193]: I am wondering if it is safe to use the configurable param here? E.g. which value of irFramesOnly will be used if it is changed from the command line, or if the parameters are taken from the CCDB (which I think should be the case at some point). Maybe we can count on the fact that inputIRFrameDataDescription will be set accordingly? But that means adding an option in the reco-workflow to do it...

File: Detectors/MUON/MCH/ROFFiltering/src/IRFrameFilter.cxx [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+using o2::InteractionRecord;
+using o2::dataformats::IRFrame;
+
+namespace o2::mch
+{
+ROFFilter createIRFrameFilter(gsl::span<const o2::dataformats::IRFrame> irframes)
+{
+  return [irframes](const ROFRecord& rof) {
+    InteractionRecord rofStart{rof.getBCData()};
+    InteractionRecord rofEnd = rofStart + rof.getBCWidth();
```
Comment (pillot) [inline Detectors/MUON/MCH/ROFFiltering/src/IRFrameFilter.cxx]: this definition of rofEnd assumes that the IRFrames are defined as [min, max[, with max excluded. Looking at the code where IRFrames are produced, e.g. CookedTrackerSpec.cxx, I have the impression that they are defined as [min, max], with max included. In which case rofEnd should be rofStart + rof.getBCWidth() - 1.

File: Detectors/MUON/MCH/ROFFiltering/src/IRFrameFilter.cxx [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    IRFrame ref(rofStart, rofEnd);
+    for (const auto& ir : irframes) {
+      // LOGP(info, "TOTO ref {} vs ir {}", ref.asString(), ir.asString());
+      auto overlap = ref.getOverlap(ir);
+      if (overlap.isValid() && !overlap.isZeroLength()) {
+        return true;
```
Comment (pillot) [inline Detectors/MUON/MCH/ROFFiltering/src/IRFrameFilter.cxx:29]: if IRFrames are defined as [min, max] with max included, then the test !overlap.isZeroLength() should be dropped too.

File: Detectors/MUON/MCH/ROFFiltering/include/MCHROFFiltering/MultiplicityFilter.h (hunk 1 above)
Comment (aphecetche) [reply Detectors/MUON/MCH/ROFFiltering/include/MCHROFFiltering/MultiplicityFilter.h]: indeed.

File: Detectors/MUON/MCH/ROFFiltering/src/IRFrameFilter.cxx (hunk 2 above)
Comment (aphecetche) [reply Detectors/MUON/MCH/ROFFiltering/src/IRFrameFilter.cxx]: TOTO is always a leftover for sure :wink:  
Forgot to grep for it before pushing apparently... Will remove it.

File: Detectors/MUON/MCH/ROFFiltering/src/fake-irframe-generator-workflow.cxx (hunk 3 above)
Comment (aphecetche) [reply Detectors/MUON/MCH/ROFFiltering/src/fake-irframe-generator-workflow.cxx:31]: correct.

File: Detectors/MUON/MCH/TimeClustering/src/TimeClusterFinderSpec.cxx (hunk 4 above)
Comment (aphecetche) [reply Detectors/MUON/MCH/TimeClustering/src/TimeClusterFinderSpec.cxx:193]: Well, can't comment on the CCDB yet (I propose we'll revisit the issue once we actually try to read the parameters from there), but for the comment line, I've used it for my tests and as far as I can tell it's working the way I expected :  

```
o2-ctf-reader-workflow --ctf-input $HOME/alice/data/2022/MAY/515056/raw/1410 --onlyDet MCH --max-tf 1 | o2-mch-fake-irframe-generator-workflow| o2-mch-reco-workflow --disable-mc --disable-root-input --configKeyValues="MCHTimeClusterizer.onlyTrackable=true;MCHDigitFilter.rejectBackground=true;MCHTimeClusterizer.peakSearchSignalOnly=true;MCHTimeClusterizer.irFramesOnly=true" --disable-root-output | o2-qc --config json://./qc-rofs.json | o2-dpl-run --run -b
```

get the `ITS/IRFrame` input, while : 

```
o2-ctf-reader-workflow --ctf-input $HOME/alice/data/2022/MAY/515056/raw/1410 --onlyDet MCH --max-tf 1 | o2-mch-fake-irframe-generator-workflow| o2-mch-reco-workflow --disable-mc --disable-root-input --configKeyValues="MCHTimeClusterizer.onlyTrackable=true;MCHDigitFilter.rejectBackground=true;MCHTimeClusterizer.peakSearchSignalOnly=true;MCHTimeClusterizer.irFramesOnly=false" --disable-root-output | o2-qc --config json://./qc-rofs.json | o2-dpl-run --run -b
```

does not.

I would prefer avoiding adding yet another option to `mch-reco-workflow` if at all possible, which I guess it is for now.

File: Detectors/MUON/MCH/ROFFiltering/src/IRFrameFilter.cxx (hunk 5 above)
Comment (aphecetche) [reply Detectors/MUON/MCH/ROFFiltering/src/IRFrameFilter.cxx]: ok

File: Detectors/MUON/MCH/ROFFiltering/src/IRFrameFilter.cxx (hunk 6 above)
Comment (aphecetche) [reply Detectors/MUON/MCH/ROFFiltering/src/IRFrameFilter.cxx:29]: ok

File: Detectors/MUON/MCH/TimeClustering/src/TimeClusterFinderSpec.cxx (hunk 4 above)
Comment (pillot) [reply Detectors/MUON/MCH/TimeClustering/src/TimeClusterFinderSpec.cxx:193]: ok

**Review by pillot — APPROVED** [review]

**Comment by aphecetche** [comment]
@shahor02 this is the PR I was talking about in the ED meeting earlier today. With the help of Philippe I can now merge my PRs w/o bothering you each time, but you are of course more than welcome to review it, if you find a moment for it ! Thanks.

**Comment by shahor02** [comment]
Looks fine to me.

**Pull request merged**